sympy/sets/sets.py

```diff
@@ -534,6 +535,69 @@ def measure(self):
         """
         return self._measure
 
+    @property
+    def kind(self):
+        """
+        The kind of a Set
+
+        Explanation
+        ===========
+
+        Any :py:class`~.Set` will have kind :py:class`~.SetKind` which is
+        parametrised by the kind of the elements of the set. For example
+        most sets are sets of numbers and will have kind
+        ``SetKind(NumberKind)``. If elements of sets are different in kind than
+        their kind will ``SetKind(UndefinedKind)``. See :py:class`~.Kind` for
+        an explantion of the kind system.
+
+        Examples
+        ========
+
+        >>> from sympy import Interval, Matrix, FiniteSet, EmptySet, ProductSet, PowerSet
+
+        >>> FiniteSet(Matrix([1, 2])).kind
+        SetKind(MatrixKind(NumberKind))
+
+        kinds of 0 and Matrix are different, kind will be
+        SetKind(UndefinedKind):
+
+        >>> Interval(1,2).kind
+        SetKind(NumberKind)
+
+        >>> EmptySet.kind
+        SetKind()
+
+        A :py:class`PowerSet` is a set of sets:
+
+        >>> PowerSet({1,2,3}).kind
+        SetKind(SetKind(NumberKind))
+
+        A :py:class`ProductSet` represents the set of tuples of elements of
+        other sets. Its kind is py:class:`TupleKind` parametrised by the kinds
+        of the elements of those sets:
+
+        >>> p = ProductSet(FiniteSet(1, 2), FiniteSet(3, 4))
+        >>> list(p)
+        [(1, 3), (2, 3), (1, 4), (2, 4)]
+        >>> p.kind
+        SetKind(TupleKind(NumberKind, NumberKind))
+
+        When all elements of the set do not have same kind, the kind
+        will be returned as ``SetKind(UndefinedKind)``:
+
+        >>> FiniteSet(0, Matrix([1, 2])).kind
+        SetKind(UndefinedKind)
+
+        >>> Interval(1,2).kind.element_kind
+        NumberKind
```

Comment:
This example could say "The kind of the elements of a set are given by the ``element_kind`` attribute of ``SetKind``.